app/durableObjects/ChatRoom.server.ts

```diff
@@ -367,35 +379,44 @@ export class ChatRoom extends Server<Env> {
 	async cleanupOldConnections() {
 		const meetingId = await this.getMeetingId()
 		if (!meetingId) log({ eventName: 'meetingIdNotFoundInCleanup' })
-		const websockets = this.ctx.getWebSockets()
+		const now = Date.now()
+		const users = await this.getUsers()
+		let removedUsers = 0
 		const connections = [...this.getConnections()]
-		log({
-			eventName: 'cleaningUpConnections',
-			meetingId,
-			connectionsFound: connections.length,
-			websocketsFound: websockets.length,
-			websocketStatuses: websockets.map((w) => w.readyState),
-		})
-		const sessionsToCleanUp = await this.getUsers()
-
-		connections.forEach((connection) =>
-			sessionsToCleanUp.delete(`session-${connection.id}`)
-		)
 
-		sessionsToCleanUp.forEach((user) => {
-			log({ eventName: 'userTimedOut', connectionId: user.id, meetingId })
-		})
+		for (const [key, user] of users) {
+			const connectionId = key.replace('session-', '')
+			const heartbeat = await this.ctx.storage.get<number>(
+				`heartbeat-${connectionId}`
+			)
+			if (heartbeat === undefined || heartbeat + 30_000 < now) {
+				removedUsers++
+				await this.ctx.storage.delete(key).catch(() => {
+					console.warn(
+						`Failed to delete session ${key} in cleanupOldConnections`
+					)
+				})
 
-		await this.ctx.storage
-			.delete([...sessionsToCleanUp.keys()])
-			.catch(() => console.warn('Failed to clean up orphaned sessions'))
+				const connection = connections.find((c) => c.id === connectionId)
+				if (connection) {
+					connection.close(1011)
+				}
+				log({ eventName: 'userTimedOut', connectionId: user.id, meetingId })
+			} else {
+				console.log(`${key} looks good`, user)
```

Comment:
This looks like it could get very noisy in production. Is that left over from development?

Reply:
Doh, yes — thanks for flagging